### ISSUE TITLE
"Strict" rule: `unnecessaryInModules` must not apply outside an actual module

### ISSUE THREAD

**Issue by depoulo** [eslint/eslint#3807]
As of now (today's `master`, but probably introduced with #1956), this rule:

[If this is modules, all "use strict" directives are unnecessary.](https://github.com/eslint/eslint/blob/master/lib/rules/strict.js#L102)

just checks if the `modules` `ecmaFeature`is enabled, instead of checking if the file tested is actually a module (by checking if the file contains an `export` statement).

This puts me in the situation that I have to turn off `strict` completely (since I'm writing ES6 `modules` in some of my (newer) (frontend) code, but not all of it (not in helpers, not in Node.js, not in old RequireJS modules, etc.).

**Comment by eslintbot** [comment]
Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by depoulo** [comment]
Version: verified with 1.1.0
**`.eslintrc`:**
```
{
    "extends": "eslint:recommended",
    "ecmaFeatures": {
        "modules": true
    },
    "env": {
        "es6": true
    },
    "rules": {
        "strict": 2
    }
}
```
**`file1.js`:**
```js
/* eslint-env browser */
import $ from 'jquery';
$.noop();
export default 'foo';
```
**`file2.js`:**
```js
/* eslint-env browser */
define(['jquery'], function ($) {
  'use strict';
  $.noop();
};
```
**`file3.js`:**
```js
/* eslint-env node */
'use strict';
module.exports = 'foo';
```

**Comment by nfroidure** [comment]
Would be nice to have this feature depending on the performance overhead since i assume it would necessitate to traverse the whole file AST.

**Comment by depoulo** [comment]
see #1956

**Comment by depoulo** [comment]
well it would be enough for me to be able to disable this behaviour via an option.

**Comment by mysticatea** [comment]
ESlint cannot decide whether or not a file is modules only from the file, so ESlint needs modules flag.

```js
import f2 from "./file2";
```

If the file2.js is loaded with a import statement, file2.js must be handled as modules.
So the strict rule is checking the modules flag.

**Comment by depoulo** [comment]
@mysticatea so what does "must be handled as modules" really mean? If there is no modules related JS code, it means almost nothing to eslint. In your example it would suffice to put `/* eslint-disable strict */` at the top of file2.js.

Alternative solution: implement `/*eslint-ecmaFeatures modules*/`

If all of this is not an option, then the current behaviour really should be made configurable (`"strict": 2, ["ignore-module"]` or whatever).

**Comment by michaelficarra** [comment]
@depoulo A module is not a JavaScript program that includes import/export statements. Just because they MAY contain import/export statements does not mean that they MUST have them. You cannot always look at a program and determine whether or not it is a module. Being a module needs to be indicated by some other means, such as being in a `<module>` tag in the browser or telling eslint to treat the program as a module.

**Comment by nzakas** [comment]
As @michaelficarra points out, you'll always need to indicate that a file is a module regardless of where you use it. In ESLint, you must set `modules: true`, and there's no other way this can work.

**Comment by depoulo** [comment]
@michaelficarra I didn't know this. Could you point me to the MDN article or W3C spec specifying the `module` tag and the fact that an ES6 module does not need to contain an export statement?

And @nzakas if I woud be able to set `module: true` as a file directive, that would be enough for me.

**Comment by michaelficarra** [comment]
@Delapouite Proof by example. Since [*ModuleBody* is optional](http://www.ecma-international.org/ecma-262/6.0/index.html#sec-modules), the empty program is a Module. The empty program does not contain an export statement.

**Comment by depoulo** [comment]
_Trying to think of a **code** example..._
Mmmh, so basically if you accepted my request, ESLint could no longer make sure nobody writes a module consisting only of the `'use strict'` pragma (where removing that pragma and leaving the file empty would fix the warning).

_How else can we fix this..._
OK so in order to use the "strict" rule without the "Illegal import declaration" error I'll need to
* create an ESLint plugin with an own environment
* in that environment, set `"ecmaFeatures": {"modules": true}` (is this even possible, @nzakas?)
* use that environment wherever I write a module

**Comment by erikmueller** [comment]
Maybe https://github.com/eslint/eslint/issues/3698 would also solve this issue as it basically let's you define a file as an es6 module?
